Firestore/Source/Remote/FSTSerializerBeta.mm

```diff
@@ -572,31 +573,95 @@ - (FieldMask)decodedFieldMask:(GCFSDocumentMask *)fieldMask {
+
+  } else {
+    FSTFail(@"Unknown transform: %d type", fieldTransform.transformation().type());
+  }
+  return proto;
+}
+
+- (GCFSArrayValue *)encodedArrayTransformElements:(const std::vector<FSTFieldValue *> &)elements {
+  GCFSArrayValue *proto = [GCFSArrayValue message];
+  NSMutableArray<GCFSValue *> *protoContents = [proto valuesArray];
+
+  for (FSTFieldValue *element : elements) {
+    GCFSValue *converted = [self encodedFieldValue:element];
+    [protoContents addObject:converted];
+  }
+  return proto;
+}
+
 - (std::vector<FieldTransform>)decodedFieldTransforms:
     (NSArray<GCFSDocumentTransform_FieldTransform *> *)protos {
   std::vector<FieldTransform> fieldTransforms;
   fieldTransforms.reserve(protos.count);
+
   for (GCFSDocumentTransform_FieldTransform *proto in protos) {
-    FSTAssert(
-        proto.setToServerValue == GCFSDocumentTransform_FieldTransform_ServerValue_RequestTime,
-        @"Unknown transform setToServerValue: %d", proto.setToServerValue);
-    fieldTransforms.emplace_back(
-        FieldPath::FromServerFormat(util::MakeStringView(proto.fieldPath)),
-        absl::make_unique<ServerTimestampTransform>(ServerTimestampTransform::Get()));
+    switch (proto.transformTypeOneOfCase) {
+      case GCFSDocumentTransform_FieldTransform_TransformType_OneOfCase_SetToServerValue: {
+        FSTAssert(
+            proto.setToServerValue == GCFSDocumentTransform_FieldTransform_ServerValue_RequestTime,
+            @"Unknown transform setToServerValue: %d", proto.setToServerValue);
+        fieldTransforms.emplace_back(
+            FieldPath::FromServerFormat(util::MakeStringView(proto.fieldPath)),
+            absl::make_unique<ServerTimestampTransform>(ServerTimestampTransform::Get()));
+        break;
+      }
+
+      case GCFSDocumentTransform_FieldTransform_TransformType_OneOfCase_AppendMissingElements: {
+        std::vector<FSTFieldValue *> elements;
+        [self decodeArrayTransformElements:proto.appendMissingElements into:elements];
+        fieldTransforms.emplace_back(
+            FieldPath::FromServerFormat(util::MakeStringView(proto.fieldPath)),
+            absl::make_unique<ArrayTransform>(TransformOperation::Type::ArrayUnion, elements));
+        break;
+      }
+
+      case GCFSDocumentTransform_FieldTransform_TransformType_OneOfCase_RemoveAllFromArray_p: {
+        std::vector<FSTFieldValue *> elements;
+        [self decodeArrayTransformElements:proto.removeAllFromArray_p into:elements];
+        fieldTransforms.emplace_back(
+            FieldPath::FromServerFormat(util::MakeStringView(proto.fieldPath)),
+            absl::make_unique<ArrayTransform>(TransformOperation::Type::ArrayRemove, elements));
+        break;
+      }
+
+      default:
+        FSTFail(@"Unknown transform: %@", proto);
+    }
   }
+
   return fieldTransforms;
 }
 
+// TODO(mikelehen): Is passing elements by reference (for use as an out param) sane / good practice?
```

Comment:
Let me know what you think...

Reply:
This is explicitly disallowed in the [Google C++ style guide](https://google.github.io/styleguide/cppguide.html#Reference_Arguments).

If you want to pass an out param, pass by pointer (i.e. pass `std::vector<FSTFieldValue *> *`).

Reply:
Changed to a pointer.  I tried a couple times to figure out how to return a `std::unique_ptr<std::vector<FSTFieldValue *>>` as that seems cleaner, but I end up getting an error I don't understand and can't seem to easily google away. ☹️

```
- (std::unique_ptr<std::vector<FSTFieldValue *>>)decodedArrayTransformElements:(GCFSArrayValue *)
    proto {
  auto elements = std::make_unique<std::vector<FSTFieldValue *>>();
  [proto.valuesArray enumerateObjectsUsingBlock:^(GCFSValue *value, NSUInteger idx, BOOL *stop) {
    // Error:(660, 5) call to implicitly-deleted copy constructor of 'const std::__1::unique_ptr<std::__1::vector<FSTFieldValue *, std::__1::allocator<FSTFieldValue *> >, std::__1::default_delete<std::__1::vector<FSTFieldValue *, std::__1::allocator<FSTFieldValue *> > > >'
    elements->push_back([self decodedFieldValue:value]);
  }];
  return elements;
}
```